src/crimson/osd/recovery_backend.cc

```diff
@@ -237,10 +237,23 @@ RecoveryBackend::scan_for_backfill(
     DEBUGDPP("querying obj:{}", pg, object);
     auto obc_manager = pg.obc_loader.get_obc_manager(
       object, /* resolve_clone = */ false);
+
+    auto found = co_await pg.obc_loader.load_and_lock(
       obc_manager, RWState::RWREAD
-    ).handle_error_interruptible(
+    ).si_then([] {
+      return true;
+    }).handle_error_interruptible(
+      crimson::ct_error::enoent::handle([](auto) {
+	return false;
+      }),
       crimson::ct_error::assert_all("unexpected error")
     );
+    if (!found) {
+      // if the object does not exist here, it must have been removed
+      // between the collection_list_partial and here.  This can happen
+      // for the first item in the range, which is usually last_backfill.
+      co_return;
```

Comment:
nit: Should we add a log line here to more easily spot cases where `scan_for_backfill` returns earlier than expected? Could help with future investigations

Reply:
Doing that now in a follow up for https://tracker.ceph.com/issues/71003